packages/vite-plugin-react-router/src/plugin.ts

```diff
@@ -0,0 +1,106 @@
+import { sep as posixSep } from 'node:path/posix'
+import { version, name } from '../package.json'
+
+const NETLIFY_FUNCTIONS_DIR = '.netlify/functions-internal'
+
+const FUNCTION_FILENAME = 'react-router-server.mjs'
+/**
+ * The chunk filename without an extension, i.e. in the Rollup config `input` format
+ */
+const FUNCTION_HANDLER_CHUNK = 'server'
+
+const FUNCTION_HANDLER_MODULE_ID = 'virtual:netlify-server'
+const RESOLVED_FUNCTION_HANDLER_MODULE_ID = `\0${FUNCTION_HANDLER_MODULE_ID}`
+
+const toPosixPath = (path: string) => path.split(sep).join(posixSep)
+
+// The virtual module that is the compiled Vite SSR entrypoint (a Netlify Function handler)
+const FUNCTION_HANDLER = /* js */ `
+import { createRequestHandler } from "@netlify/vite-plugin-react-router";
+import * as build from "virtual:react-router/server-build";
+export default createRequestHandler({
+  build,
+  getLoadContext: async (_req, ctx) => ctx,
+});
+`
+
+// This is written to the functions directory. It just re-exports
+// the compiled entrypoint, along with Netlify function config.
+function generateNetlifyFunction(handlerPath: string) {
+  return /* js */ `
+    export { default } from "${handlerPath}";
+
+    export const config = {
+      name: "React Router server handler",
+      generator: "${name}@${version}",
+      path: "/*",
+      preferStatic: true,
+    };
+    `
+}
+
+export function netlifyPlugin(): Plugin {
+  let resolvedConfig: ResolvedConfig
+  let currentCommand: string
+  let isSsr: boolean | undefined
+  return {
+    name: 'vite-plugin-react-router-netlify-functions',
+    config(config, { command, isSsrBuild }) {
+      currentCommand = command
+      isSsr = isSsrBuild
+      if (command === 'build') {
+        if (isSsrBuild) {
+          // We need to add an extra SSR entrypoint, as we need to compile
+          // the server entrypoint too. This is because it uses virtual
+          // modules.
+          // NOTE: the below is making various assumptions about the React Router Vite plugin's
+          // implementation details:
+          // https://github.com/remix-run/remix/blob/cc65962b1a96d1e134336aa9620ef1dad7c5efb1/packages/remix-dev/vite/plugin.ts#L1149-L1168
+          // TODO(serhalp) Stop making these assumptions or assert them explictly.
+          // TODO(serhalp) Unless I'm misunderstanding something, we should only need to *replace*
+          // the default React Router Vite SSR entrypoint, not add an additional one.
+          if (typeof config.build?.rollupOptions?.input === 'string') {
+            config.build.rollupOptions.input = {
+              [FUNCTION_HANDLER_CHUNK]: FUNCTION_HANDLER_MODULE_ID,
+              index: config.build.rollupOptions.input,
+            }
+            if (config.build.rollupOptions.output && !Array.isArray(config.build.rollupOptions.output)) {
+              config.build.rollupOptions.output.entryFileNames = '[name].js'
+            }
+          }
+        }
```

Comment:
I think we can probably simplify this mess a lot. Avoid scope creep, but maybe timebox 10 mins on rewriting this part from scratch. Remix's own template just did this and it worked: https://github.com/remix-run/react-router-templates/blob/ff303a5bd4c742104c601dc5eda4f825d10739b8/netlify/vite.config.ts#L8-L14.

Reply:
Let's do it!

Reply:
Woo, this sorta just worked too. Ultimately it ends up being roughly the same number of lines, but I think it's easier to understand and no longer requires a bunch of TODOs and NOTEs. :)

We should also be able to backport this refactor to the Remix plugins later.